aten/src/ATen/core/dispatch/Dispatcher.cpp

```diff
@@ -1,8 +1,87 @@
 #include <ATen/core/dispatch/Dispatcher.h>
 
 namespace c10 {
+
+namespace detail {
+class RegistrationListenerList final {
+public:
+  void addListener(std::unique_ptr<OpRegistrationListener> listener) {
+    listeners_.push_back(std::move(listener));
+  }
+
+  void callOnOperatorRegistered(const OperatorHandle& op) {
+    for (auto& listener : listeners_) {
+      listener->onOperatorRegistered(op);
+    }
+  }
+
+  void callOnOperatorDeregistered(const OperatorHandle& op) {
+    for (auto& listener : listeners_) {
+      listener->onOperatorDeregistered(op);
+    }
+  }
+private:
+  std::vector<std::unique_ptr<OpRegistrationListener>> listeners_;
+};
+}
+
+Dispatcher::Dispatcher()
+: operators_()
+, listeners_(guts::make_unique<detail::RegistrationListenerList>())
+, mutex_() {}
+
+Dispatcher::~Dispatcher() {}
+
 C10_EXPORT Dispatcher& Dispatcher::singleton() {
   static Dispatcher _singleton;
   return _singleton;
 }
+
+OperatorHandle Dispatcher::registerSchema(FunctionSchema schema) {
+  // we need a lock to avoid concurrent writes
+  std::lock_guard<std::mutex> lock(mutex_);
+
+  operators_.emplace_back(std::move(schema));
+  auto op = OperatorHandle(--operators_.end());
```

Comment:
nit: rbegin?

Reply:
That would either require using reverse_iterator's everywhere, or I would have to cast the reverse_iterator back to a normal iterator (which needs the offset in the cast because just casting a reverse_iterator to an iterator without adding an offset results in it pointing to a different element).